client/src/procedures/primitives.rs

```diff
@@ -0,0 +1,1008 @@
+            target: OutputKey::random(),
+            is_temp: true,
+        };
+        let tag = InterimProduct {
+            target: OutputKey::random(),
+            is_temp: true,
+        };
+        AeadEncrypt {
+            associated_data,
+            plaintext,
+            nonce,
+            key,
+            ciphertext,
+            tag,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn store_ciphertext(mut self, key: OutputKey) -> Self {
+        self.ciphertext = InterimProduct {
+            target: key,
+            is_temp: false,
+        };
+        self
+    }
+
+    pub fn store_tag(mut self, key: OutputKey) -> Self {
+        self.tag = InterimProduct {
+            target: key,
+            is_temp: false,
+        };
+        self
+    }
+}
+
+impl<T> SourceInfo for AeadEncrypt<T> {
+    fn source_location(&self) -> &Location {
+        &self.key
+    }
+    fn source_location_mut(&mut self) -> &mut Location {
+        &mut self.key
+    }
+}
+
+impl<T: Aead> ProcedureStep for AeadEncrypt<T> {
+    fn execute<R: Runner>(self, runner: &mut R, state: &mut State) -> Result<(), anyhow::Error> {
+        let AeadEncrypt {
+            associated_data,
+            plaintext,
+            nonce,
+            key,
+            ciphertext,
+            tag,
+            ..
+        } = self;
+        let plaintext = match plaintext {
+            InputData::Value(ref v) => v,
+            InputData::Key(key) => {
+                let data = state.get_data(&key)?;
+                data.as_ref()
+            }
+        };
+        let nonce = match nonce {
+            InputData::Value(ref v) => v,
+            InputData::Key(key) => {
+                let data = state.get_data(&key)?;
+                data.as_ref()
+            }
+        };
+        let ad = match associated_data {
+            InputData::Value(ref v) => v,
+            InputData::Key(key) => {
+                let data = state.get_data(&key)?;
+                data.as_ref()
+            }
+        };
+
+        let mut digested = Vec::new();
+        let mut t = Tag::<T>::default();
+
+        let f = |key: GuardedVec<u8>| {
+            T::try_encrypt(&*key.borrow(), nonce, ad, plaintext, &mut digested, &mut t)
+                .map_err(engine::Error::CryptoError)
+        };
+
+        runner.get_guard(&key, f).map_err(|e| anyhow::anyhow!(e))?;
+        state.insert_data(ciphertext.target, digested.into(), ciphertext.is_temp);
+        state.insert_data(tag.target, Vec::from(&*t).into(), tag.is_temp);
+        Ok(())
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct AeadDecrypt<T> {
+    associated_data: InputData<Vec<u8>>,
+    ciphertext: InputData<Vec<u8>>,
+    tag: InputData<Vec<u8>>,
+    nonce: InputData<Vec<u8>>,
+    key: Location,
+    plaintext: InterimProduct<OutputKey>,
+    _marker: PhantomData<T>,
+}
+
+impl<T> AeadDecrypt<T> {
+    pub fn new(
+        key: Location,
```

Comment:
maybe fewer arguments to the function call? Might there be a possibility to combine them into a parameters struct, that can be reused among individual calls?

Reply:
I do kind of agree here, it would clean up the code a little to make a decryption input type or something along those lines.